public/components/register_model/model_task_manager.ts

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Observable, timer, BehaviorSubject } from 'rxjs';
+import { takeWhile, switchMap } from 'rxjs/operators';
+import { APIProvider } from '../../apis/api_provider';
+import { TaskGetOneResponse } from '../../apis/task';
+
+interface TaskQueryOptions {
+  taskId: string;
+  onUpdate?: (status: TaskGetOneResponse) => void;
+  onError?: (err: Error) => void;
+  onComplete?: (status: TaskGetOneResponse) => void;
+}
+
+export class ModelTaskManager {
+  /**
+   * The model download tasks which are still running in BE
+   */
+  tasks = new BehaviorSubject<Map<string, Observable<TaskGetOneResponse>>>(new Map());
+
+  constructor() {}
+
+  remove(taskId: string) {
+    this.tasks.getValue().delete(taskId);
+    this.tasks.next(this.tasks.getValue());
+  }
+
+  query(options: TaskQueryOptions) {
+    if (!this.tasks.getValue().has(options.taskId)) {
+      const observable = timer(0, 2000)
+        .pipe(switchMap((_) => APIProvider.getAPI('task').getOne(options.taskId)))
+        // TODO: should it also check res.state?
+        // The intention here is to stop polling once a model is created
+        .pipe(takeWhile((res) => !Boolean(res.model_id), true));
+
+      observable.subscribe({
+        next: (res) => {
+          if (options.onUpdate && !res.error) {
+            options.onUpdate(res);
+          }
+          // Model download task is still running if
+          // 1. model id doesn't exist
+          // 2. the current task is running fine without error
+          if (!res.model_id && !res.error) {
+            this.tasks.next(this.tasks.getValue().set(options.taskId, observable));
```

Comment:
Could we append this tasks outside? From the code I think this setter will be called after first get task detail API response, it may cause race condition.

Reply:
Updated so that it only set the task once.

My intention was only adding the running task to `this.task`, if we know a task is completed or has error in the first call, then we don't add it to `this.tasks`. So that we can know the current running task count by subscribe to `this.tasks`.

Btw, what's the race condition you mentioned?

Reply:
I'm misunderstanding here. What I originally wanted to ask is that there is a time gap from the user calls the query method to the tasks add the task to the map. If you access the tasks map within this time gap, user won't get the corresponding task. Do we need to concern about this problem?

Reply:
`this.tasks` can be converted to observable, see `this.getTasks$`, user can subscribe to the observable to get the changes

Reply:
Sure. Is that means the changes notified after first get task detail API called? I think it's OK if these is intention.

Reply:
Yes, I was thinking of only pushing it to `tasks` when we know the task is in progress from the first API response. So `this.tasks` represents the tasks which are in progress.